src/open_inwoner/utils/migration_operations.py

```
@@ -1,59 +1,339 @@
+
+    logger = logging.getLogger(__name__)
+
+    # Set defaults
+    if allowed_node_types is None:
+        allowed_node_types = [NodeType.PARAGRAPH]
+
+    if allowed_mark_types is None:
+        allowed_mark_types = [
+            MarkType.STRONG,
+            MarkType.ITALIC,
+            MarkType.LINK,
+            MarkType.UNDERLINE,
+        ]
+
+    # Create Prosemirror config
+    config = ProsemirrorConfig(
+        allowed_node_types=allowed_node_types,
+        allowed_mark_types=allowed_mark_types,
+    )
+
+    # Get model
+    Model = apps.get_model(app_label, model_name)
+
+    # Get instances to process
+    if use_singleton:
+        instances = [Model.objects.first()] if Model.objects.exists() else []
+    else:
+        instances = Model.objects.all()
+
+    # Process each instance
+    for instance in instances:
+        if not instance:
+            continue
+
+        # Get original content
+        old_content = getattr(instance, f"{field_name}_tmp")
+
+        # Check if old_content has actual text (not just empty tags)
+        if not _has_text_content(old_content):
+            logger.info(
+                "Skipping %s.%s for instance %s: no text content",
+                model_name,
+                field_name,
+                instance.pk if hasattr(instance, "pk") else instance,
+            )
+            setattr(instance, field_name, None)
+            instance.save()
+            continue
+
+        # Clean empty tags before parsing (prevents IndexError)
+        content = _clean_empty_html_tags(old_content)
+
+        # After cleaning, double-check again if anything is left
+        if not content or not content.strip():
+            setattr(instance, field_name, None)
+            instance.save()
+            continue
+
+        try:
+            # Convert markdown to HTML if needed
+            if convert_markdown:
+                content = markdown.markdown(content, extensions=["extra"])
```

Final thought on this: markdown and HTML are mutually exclusive right? And if so, is there a chance the earlier cleaning operations will generate errors (because they're working on markdown, not HTML), or does thew built-in HTML parser silently ignore everything in that case?

---

Good catch! They're not exclusive (markdown can contain inline HTML) and the cleaning operations work on HTML not markdown, but there could be bugs: e.g. if the cleaning operations modify/strip inline HTML and the markdown conversion is feeded corrupted markdown. I'll fix it, thanks.